### ISSUE TITLE
[Bug? or Documentation outdated]  Type conversion for int --> month not working

### ISSUE THREAD

**Issue by jeffward01** [Qowaiv/Qowaiv#288]
Hello,

I am working on some basic things such as comparing the current month from `DateTime.UtcNow.Month` to `Month` in the `Qowaiv` library. 

I am unsure how to do this cleanly using the library, so I have referred to the docs.  In the docs I see under **Month**

```csharp
    [Fact]
    public void ShouldConvertToMonthInt()
    {
        Month feb = Month.Parse("February", CultureInfo.CurrentCulture);
        Month may = Month.May;
       
        // This below will not compile
        Month dec = 12;

        Assert.Equal(dec, Month.December);
    }
}
```

However, this does not compile due to a type converter not implemented.

Instead, I must do this:

```csharp
    [Fact]
    public void ShouldConvertToMonthInt()
    {
        Month feb = Month.Parse("February", CultureInfo.CurrentCulture);
        Month may = Month.May;
        Month dec = (Month)12;

        Assert.Equal(dec, Month.December);
    }
```

Questions:

1.)  Are the docs outdated?
2) Are there any recommended ways to compare `DateTime` with `Year` or `Month`

Thanks

---

![image](https://user-images.githubusercontent.com/12589359/211222839-b8350b0d-197e-45fb-8659-e57cd5bb030a.png)
[Link to docs](https://github.com/Qowaiv/Qowaiv#month)

**Comment by Corniel** [comment]
Thanks for spotting an issue in the documentation. Those implicit castings haven been dropped (see #212) when introducing version 6.

To create a `Month` from an int you can use `Month.Create(int)` or indeed an explicit cast.

I've considered adding a `GetMonth()` extension on a `DateTime` and ``DateOnly` (and on the date types in Qowaiv). I still can do that. An other extesion could be to implement `IEquality<T>` on month (and year) for mutiple date/date time types, and extensions on `DateTime`. That can both be `Equals(Month month)` or, `HasMonth(Month month)`. From a domain-driven perspective you could argue that the second one has the better naming.

Can you tell me something about your use case?

**Comment by jeffward01** [comment]
> I've considered adding a `GetMonth()` extension on a `DateTime` and ``DateOnly`(and on the date types in Qowaiv). I still can do that. An other extesion could be to implement`IEquality` on month (and year) for mutiple date/date time types, and extensions on `DateTime`. That can both be `Equals(Month month)` or, `HasMonth(Month month)`. From a domain-driven perspective you could argue that the second one has the better naming.

I think that would be very cool and useful.  Most of the time when people work with value objects, they are transforming `string` to `FirstName` as an example.  For this referenced issue with `Month`, they would be trasnforming `int` or `DateTime` to `Month`.

> Can you tell me something about your use case?

Of course!

For this particular use case, I am using it for _Credit Card expiration dates_.

To be concise, I wanted a way to determine if a credit card was expired.  First by checking the year, then the year and month.

`var expirationYear = Year.Create(2010);`

` var shouldBeExpired = DateTime.UtcNow > expirationYear`



or to be more related to the ticket, I also use this for events (like appointments or something)

```csharp

public void CreateEvent(DateTime dateTimeFromApiJson)
{
   var year =  dateTimeFromApiJson.GetYear();
   var month =  dateTimeFromApiJson.GetMonth);   
  // etc.
  // the goal is to work with something a bit more flexible than DateTime, similar to DateOnly, but more modular
}

```


Here is a value object I implemented for credit card expiration, personally, I dislike how usage of `Qowaiv` requires the use of `(int)`, in my mind, Casting is something that is more advanced, like casting to an implementation of a certain interface, rather than just "standard use" of the library


```csharp

public class CreditCardExpirationDate : ValueObject<CreditCardExpirationDate>,
    IValueState
{
    private CreditCardExpirationDate(
        Month month,
        Year year)
    {
        this.ExpirationMonth = month;
        this.ExpirationYear = year;
    }

    private CreditCardExpirationDate()
    {
        this.ExpirationMonth = Month.Empty;
        this.ExpirationYear = Year.Empty;
    }

    public Month ExpirationMonth { get; }
    public Year ExpirationYear { get; }

    /// <inheritdoc />
    public bool IsEmpty => !this.HasValue;

    /// <inheritdoc />
    public bool HasValue => this.ExpirationYear != Year.Empty && this.ExpirationMonth != Month.Empty;

    public bool IsExpired(DateTime? expirationDate = null)
    {
        DateTime evalExpirationDate = expirationDate ?? DateTime.UtcNow;

        return this._isCardExpired(evalExpirationDate, this.ExpirationMonth, this.ExpirationYear);
    }

    public static CreditCardExpirationDate From(
        int month,
        int year)
    {
        if (!IsValidMonth(month))
        {
            throw new InvalidDataException("The credit card expiration month is invalid.");
        }

        if (!IsValidYear(year))
        {
            throw new InvalidDataException("The credit card expiration year is invalid.");
        }

        return new CreditCardExpirationDate(Month.Create(month), Year.Create(year));
    }

    /// <inheritdoc />
    protected override bool EqualsCore(CreditCardExpirationDate other)
    {
        return this.ExpirationMonth == other.ExpirationMonth && this.ExpirationYear == other.ExpirationYear;
    }

    /// <inheritdoc />
    protected override int GetHashCodeCore()
    {
        int hashCode = this.ExpirationYear.GetHashCode();
        hashCode = (hashCode * 397) ^ this.IsEmpty.GetHashCode();
        hashCode = (hashCode * 397) ^ this.HasValue.GetHashCode();
        hashCode = (hashCode * 397) ^ this.ExpirationMonth.GetHashCode();

        return hashCode;
    }

    public static bool IsValidMonth(int month)
    {
        return month is >= 1 and <= 12;
    }

    public static bool IsValidYear(int year)
    {
        return year >= 0;
    }
    
 
    private bool _isCardExpired(
        DateTime evalDate,
        Month month,
        Year year)
    {
        if (evalDate.Year > (int)year)
        {
            return true;
        }

        if (evalDate.Month > (int)month)
        {
            return true;
        }

        return false;
    }
}
```

Note: the class `ValueObject<T>` is from [CSharpFunctionalExtensions](https://github.com/vkhorikov/CSharpFunctionalExtensions)

----

Basically, it would be nice if `Year` and `Month` were a bit more easier to work with when comparing to `DateTime.Utc`, or moving back and forth between ValueObjects and `DateTime` structs,

----

I hope this helped and made sense, thanks for the AWESOME library! you guys are awesome

**Comment by Corniel** [comment]
I would argue that what you need here is a  `YearMonth`. A SVO that is on the to-do list as I need it myself. That aside, both Month and Year have a TryCreate.

So you could create a TryCreate yourself:

``` c#
static CreditCardExpirationDate? TryCreate(int year, int month)
    => Year.TryCreate(year, out var y) && Month.TryCreate(month, out var m)
    ? new(y, m)
    : null;
```

Your point about the comparison options between `Year` and `Month` and different date related types is still valid. The Qowaiv API needs to be extended. I'm considering: `*.InMonth(Month month)`, `*.WithinnMonth(Month month)` and `*.HasMonth(Month month)` (and the year equivalents too, obviously). What do you think?

## Off-topic
For UtcNow related constraints, you might want to use [Qowaiv.Clock](https://github.com/Qowaiv/Qowaiv#qowaiv-clock)

For Hashing Value Objects, you might want to use [Qowaiv.Hash]( https://github.com/Qowaiv/Qowaiv#hashing)

**Comment by Corniel** [comment]
@jeffward01 I shipped a new version including the _date_`.IsIn(Month)` and  _date_`.IsIn(Year)` features. I expect to work on a YearMonth Svo next week or so. I'll close this issue. If you think that is not correct. Feel free to respond, and I'll reopen it.
